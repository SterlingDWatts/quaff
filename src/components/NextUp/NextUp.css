.NextUp {
  text-align: center;
  position: relative;
  overflow: hidden;
}

.NextUp a {
  color: white;
  text-decoration: none;
}

.NextUp a:hover .NextUp__header,
.NextUp a:active .NextUp__header,
.NextUp a:focus .NextUp__header {
  background-color: #e30425;
  color: white;
}

.NextUp__header {
  color: #e30425;
  background-color: white;
  line-height: 1;
  font-family: "Lekton", sans-serif;
  font-size: 16px;
  font-weight: 600;
  position: absolute;
  padding: 8px 16px;
  border-bottom-right-radius: 8px;
  z-index: 5;
}

.NextUp__container {
  height: 150px;
  display: flex;
  flex-direction: column;
  justify-content: center;
  background-color: white;
  background-position: center;
  background-size: cover;
  font-size: 48px;
  font-weight: 600;
  text-shadow: 1px 1px rgba(0, 0, 0, 0.5);
  line-height: 1;
  color: white;
}

.NextUp__learn-loading .NextUp__container,
.NextUp__study-loading .NextUp__container {
  border: 2px solid #f2f4f5;
}

.NextUp__learn-loading .NextUp__header,
.NextUp__study-loading .NextUp__header {
  border-bottom: 2px solid #f2f4f5;
  border-right: 2px solid #f2f4f5;
}

.NextUp--loading {
  height: 48px;
  width: 175px;
  background-color: rgb(242, 244, 245);
  margin: auto;
}

@keyframes load {
  from {
    left: -150px;
  }
  to {
    left: 100%;
  }
}

.NextUp__placeholder::before {
  content: " ";
  display: block;
  width: 150px;
  position: absolute;
  top: 0;
  bottom: 0;
  left: -150px;
  background: linear-gradient(
    to top right,
    transparent 0%,
    white 50%,
    transparent 100%
  );
  animation: load 1s cubic-bezier(0.4, 0, 0.2, 1) infinite;
  z-index: 6;
}

.NextUp__header-loading {
  width: 56px;
  height: 16px;
  background-color: rgb(242, 244, 245);
}

.NextUp__container:hover,
.NextUp__container:active,
.NextUp__container:focus {
  box-shadow: 0 3px 6px rgba(0, 0, 0, 0.16), 0 3px 6px rgba(0, 0, 0, 0.23);
  color: rgba(250, 250, 250, 0.5);
}

.NextUp__container--locked {
  filter: grayscale(100%);
}

@media screen and (min-width: 500px) {
  .NextUp {
    width: 500px;
    margin: 0 auto;
  }

  .NextUp__container {
    margin-bottom: 16px;
  }
}
