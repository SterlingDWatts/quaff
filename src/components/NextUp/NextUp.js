import React, { Component } from "react";
import { Link } from "react-router-dom";
import QuizListContext from "../../contexts/QuizListContext";
import "./NextUp.css";

class NextUp extends Component {
  static contextType = QuizListContext;

  renderNextModule = () => {
    let nextModule = this.context.quizList.findIndex(
      (quiz) => quiz.unlocked === false
    );
    nextModule = nextModule ? nextModule - 1 : this.context.quizList.length - 1;
    nextModule = this.context.quizList[nextModule];
    return (
      <NextUpItem
        id={nextModule.id}
        label={nextModule.name}
        type={"Learn"}
        picture={nextModule.picture}
      />
    );
  };

  renderStudyTopics = () => {
    let topicList = this.context.topicList.sort(
      (a, b) => a.mastery - b.mastery
    );
    topicList = topicList.slice(0, 3);
    return topicList.map((topic) => {
      return (
        <NextUpItem
          key={topic.id}
          id={topic.id}
          label={topic.name}
          type={"Study"}
          picture={topic.picture}
        />
      );
    });
  };

  renderLoading = () => {
    return (
      <>
        <div className="NextUp__learn-loading">
          <div className="NextUp__header">Learn</div>
          <div className="NextUp__container">Loading</div>
        </div>
        <div className="NextUp__study-loading">
          <div className="NextUp__header">Learn</div>
          <div className="NextUp__container">Loading</div>
        </div>
        <div className="NextUp__study-loading">
          <div className="NextUp__header">Learn</div>
          <div className="NextUp__container">Loading</div>
        </div>
        <div className="NextUp__study-loading">
          <div className="NextUp__header">Learn</div>
          <div className="NextUp__container">Loading</div>
        </div>
      </>
    );
  };

  render() {
    const { error, topicList } = this.context;
    let content;
    if (error) {
      content = <p className="NextUp--error">There was an error</p>;
    } else if (topicList.length < 1) {
      content = this.renderLoading();
    } else {
      content = (
        <>
          {this.renderNextModule()}
          {this.renderStudyTopics()}
        </>
      );
    }
    return <div className="NextUp">{content}</div>;
  }
}

function NextUpItem({ id, label, picture, type }) {
  return (
    <Link to={`/${type.toLowerCase()}/${id}`}>
      <div className="NextUp__header">{type}</div>
      <div
        className="NextUp__container"
        style={{ backgroundImage: `url(${picture})` }}
      >
        {label}
      </div>
    </Link>
  );
}

export default NextUp;
