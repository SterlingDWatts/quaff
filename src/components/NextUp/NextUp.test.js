import React from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { BrowserRouter } from "react-router-dom";
import NextUp from "./NextUp";

describe("NextUp Component", () => {
  it("renders without crashing", () => {
    const div = document.createElement("div");
    ReactDOM.render(
      <BrowserRouter>
        <NextUp />
      </BrowserRouter>,
      div
    );
    ReactDOM.unmountComponentAtNode(div);
  });

  it("renders the UI as expected", () => {
    const tree = renderer
      .create(
        <BrowserRouter>
          <NextUp />
        </BrowserRouter>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
