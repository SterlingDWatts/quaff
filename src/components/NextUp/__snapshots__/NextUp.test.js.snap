// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NextUp Component renders the UI as expected 1`] = `
<div
  className="NextUp"
>
  <div
    className="NextUp__learn-loading"
  >
    <div
      className="NextUp__header"
    >
      Learn
    </div>
    <div
      className="NextUp__container"
    >
      Loading
    </div>
  </div>
  <div
    className="NextUp__study-loading"
  >
    <div
      className="NextUp__header"
    >
      Learn
    </div>
    <div
      className="NextUp__container"
    >
      Loading
    </div>
  </div>
  <div
    className="NextUp__study-loading"
  >
    <div
      className="NextUp__header"
    >
      Learn
    </div>
    <div
      className="NextUp__container"
    >
      Loading
    </div>
  </div>
  <div
    className="NextUp__study-loading"
  >
    <div
      className="NextUp__header"
    >
      Learn
    </div>
    <div
      className="NextUp__container"
    >
      Loading
    </div>
  </div>
</div>
`;
